Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 Host: ip-10-20-0-213.ca-central-1.compute.internal
 GCC:  gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Git: 2.9.5,  CMake: 3.22.1,  cUrl: 7.67.0,  node.js: v16.13.1,  npm: 8.1.2
Sha: bad7c57c478686a9182e31581a12b57a5354ba2a
Containerized:False
Number of Error(s): 58
cc: error: /home/centos/smoketest/PR-15730/HPCC-Platform/common/remote/hooks/git/libgit2/src/netops.c: No such file or directory
cc: fatal error: no input files
make[2]: *** [common/remote/hooks/git/libgit2/src/CMakeFiles/git2internal.dir/build.make:1084: common/remote/hooks/git/libgit2/src/CMakeFiles/git2internal.dir/netops.c.o] Error 1
/home/centos/smoketest/PR-15730/HPCC-Platform/common/remote/hooks/git/libgit2/src/net.c:15:10: fatal error: http_parser.h: No such file or directory
make[2]: *** [common/remote/hooks/git/libgit2/src/CMakeFiles/git2internal.dir/build.make:1070: common/remote/hooks/git/libgit2/src/CMakeFiles/git2internal.dir/net.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:9318: common/remote/hooks/git/libgit2/src/CMakeFiles/git2internal.dir/all] Error 2
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
c++: error: /home/centos/smoketest/PR-15730/HPCC-Platform/plugins/nlp/nlp-engine/cs/libconsh/st_gen.cpp: No such file or directory
c++: fatal error: no input files
make[2]: *** [plugins/nlp/nlp-engine/cs/libconsh/CMakeFiles/consh.dir/build.make:370: plugins/nlp/nlp-engine/cs/libconsh/CMakeFiles/consh.dir/st_gen.cpp.o] Error 1
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
cannot fetch initial working directory: No such file or directory at /usr/share/autoconf/Autom4te/General.pm line 196.
aclocal: error: echo failed with exit status: 1
make[2]: *** [plugins/cassandra/CMakeFiles/libuv.dir/build.make:73: plugins/cassandra/.libs/libuv.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:15575: plugins/cassandra/CMakeFiles/libuv.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:15057: plugins/nlp/nlp-engine/cs/libconsh/CMakeFiles/consh.dir/all] Error 2
CMake Error: File /home/centos/smoketest/PR-15730/HPCC-Platform/plugins/couchbase/libcouchbase/cmake/config-cmake.h.in does not exist.
CMake Error at cmake/Modules/GenerateConfigDotH.cmake:47 (CONFIGURE_FILE):
-- Configuring incomplete, errors occurred!
make[2]: *** [plugins/couchbase/CMakeFiles/libcouchbase-build.dir/build.make:74: plugins/couchbase/libcouchbase-build/lib/libcouchbase.so] Error 1
make[1]: *** [CMakeFiles/Makefile2:15891: plugins/couchbase/CMakeFiles/libcouchbase-build.dir/all] Error 2
CMake Error: The source directory /home/centos/smoketest/PR-15730/HPCC-Platform/system/logaccess/ElasticStack/elasticlient does not exist.
gmake[4]: *** [CMakeFiles/Makefile2:309: cmake_check_build_system] Error 1
gmake[3]: *** [Makefile:169: jsoncpp_lib] Error 2
make[2]: *** [system/logaccess/ElasticStack/CMakeFiles/elasticlient-build.dir/build.make:86: system/logaccess/ElasticStack/elasticlient-build-prefix/src/elasticlient-build-stamp/elasticlient-build-build] Error 2
make[1]: *** [CMakeFiles/Makefile2:7394: system/logaccess/ElasticStack/CMakeFiles/elasticlient-build.dir/all] Error 2
make[2]: *** [CMakeFiles/generate-boost-regex.dir/build.make:101: generate-boost-regex-prefix/src/generate-boost-regex-stamp/generate-boost-regex-download] Error 1
make[1]: *** [CMakeFiles/Makefile2:7097: CMakeFiles/generate-boost-regex.dir/all] Error 2
/home/centos/smoketest/PR-15730/build/system/aws/aws-sdk-cpp/.deps/build/src/AwsCCommon/includ
 ... ( comment is too long, 383 lines, 31895 bytes, truncated) ... 
[Build error:](https://gist.github.com/ad9e0b3e82ee947674a632b60bcfa4d8)
Build log [Build log](https://gist.github.com/HPCCSmoketest/ad9e0b3e82ee947674a632b60bcfa4d8#file-result-22-02-03-18-39-24-log)

```

--- a/dali/base/dadfs.cpp
+++ b/dali/base/dadfs.cpp
@@ -177,7 +177,7 @@ static IPropertyTree *getEmptyAttr()
     return createPTree("Attr");
 }
 
-extern da_decl double calcFileCost(const char * cluster, double sizeGB, double fileAgeDays, __int64 numDiskWrites, __int64 numDiskReads)
+extern da_decl void calcFileCost(const char * cluster, double sizeGB, double fileAgeDays, __int64 numDiskWrites, __int64 numDiskReads, double & atRestCost, double & accessCost)
 {
     Owned<IPropertyTree> plane = getStoragePlane(cluster);
     Owned<IPropertyTree> global;
@@ -193,13 +193,25 @@ extern da_decl double calcFileCost(const char * cluster, double sizeGB, double f
         costPT = global->queryPropTree("cost");
     }
     if (costPT==nullptr)
-        return 0.0;
+    {
+        atRestCost = 0.0;
+        accessCost = 0.0;
+        return;
+    }
     constexpr int accessPriceScalingFactor = 10000; // read/write pricing based on 10,000 operations
     double atRestPrice = costPT->getPropReal("@storageAtRest", 0.0);
     double readPrice =  costPT->getPropReal("@storageReads", 0.0);
     double writePrice =  costPT->getPropReal("@storageWrites", 0.0);
     double storageCostDaily = atRestPrice * 12 / 365;
-    return (storageCostDaily * sizeGB * fileAgeDays) + (readPrice * numDiskReads / accessPriceScalingFactor) + (writePrice * numDiskWrites / accessPriceScalingFactor);
+    atRestCost = storageCostDaily * sizeGB * fileAgeDays;
+    accessCost = (readPrice * numDiskReads / accessPriceScalingFactor) + (writePrice * numDiskWrites / accessPriceScalingFactor);
+}
+
+extern da_decl double calcFileAccessCost(const char * cluster, __int64 numDiskWrites, __int64 numDiskReads)
+{
+    double atRestCost, accessCost;
+    calcFileCost(cluster, 0, 0, numDiskWrites, numDiskReads, atRestCost, accessCost);
+    return accessCost;
 }
 
 RemoteFilename &constructPartFilename(IGroup *grp,unsigned partno,unsigned partmax,const char *name,const char *partmask,const char *partdir,unsigned copy,ClusterPartDiskMapSpec &mspec,RemoteFilename &rfn)
@@ -4787,7 +4799,7 @@ public:
         else
             return false;
     }
-    virtual double getCost(const char * cluster) override
+    virtual void getCost(const char * cluster, double & atRestCost, double & accessCost) override
     {
         CDateTime dt;
         getModificationTime(dt);
@@ -4800,19 +4812,21 @@ public:
             numDiskWrites = attrs->getPropInt64("@numDiskWrites");
             numDiskReads = attrs->getPropInt64("@numDiskReads");
         }
-
         if (isEmptyString(cluster))
         {
             StringArray clusterNames;
             unsigned countClusters = getClusterNames(clusterNames);
-            double totalCost = 0.0;
             for (unsigned i = 0; i < countClusters; i++)
-                totalCost += calcFileCost(clusterNames[i], sizeGB, fileAgeDays, numDiskWrites, numDiskReads);
-            return totalCost;
+            {
+                double tmpAtRestcost, tmpAccessCost;
+                calcFileCost(clusterNames[i], sizeGB, fileAgeDays, numDiskWrites, numDiskReads, tmpAtRestcost, tmpAccessCost);
+                atRestCost += tmpAtRestcost;
+                accessCost += tmpAccessCost;
+            }
         }
         else
         {
-            return calcFileCost(cluster, sizeGB, fileAgeDays, numDiskWrites, numDiskReads);
+            calcFileCost(cluster, sizeGB, fileAgeDays, numDiskWrites, numDiskReads, atRestCost, accessCost);
         }
     }
 };
@@ -6783,16 +6797,17 @@ public:
         return false;
     }
 
-    virtual double getCost(const char * cluster) override
+    virtual void getCost(const char * cluster, double & atRestCost, double & accessCost) override
     {
-        double totalCost = 0.0;
         CriticalBlock block (sect);
         ForEachItemIn(i,subfiles)
         {
+            double tmpAtRestcost, tmpAccessCost;
             IDistributedFile &f = subfiles.item(i);
-            totalCost += f.getCost(cluster);
+            f.getCost(cluster, tmpAtRestcost, tmpAccessCost);
+            atRestCost += tmpAtRestcost;
+            accessCost += tmpAccessCost;
         }
-        return totalCost;
     }
 };
 
@@ -12818,7 +12833,7 @@ IDFProtectedIterator *CDistributedFileDirectory::lookupProtectedFiles(const char
 const char* DFUQResultFieldNames[] = { "@name", "@description", "@group", "@kind", "@modified", "@job", "@owner",
     "@DFUSFrecordCount", "@recordCount", "@recordSize", "@DFUSFsize", "@size", "@workunit", "@DFUSFcluster", "@numsubfiles",
     "@accessed", "@numparts", "@compressedSize", "@directory", "@partmask", "@superowners", "@persistent", "@protect", "@compressed",
-    "@cost", "@numDiskReads", "@numDiskWrites" };
+    "@cost", "@numDiskReads", "@numDiskWrites", "@atRestCost", "@accessCost" };
 
 extern da_decl const char* getDFUQResultFieldName(DFUQResultField feild)
 {
@@ -12905,8 +12920,11 @@ IPropertyTreeIterator *deserializeFileAttrIterator(MemoryBuffer& mb, unsigned nu
             double sizeGB = sizeDiskSize / ((double)1024 * 1024 * 1024);
             __int64 numDiskWrites = file->getPropInt64(getDFUQResultFieldName(DFUQRFnumDiskReads), 0);
             __int64 numDiskReads = file->getPropInt64(getDFUQResultFieldName(DFUQRFnumDiskWrites), 0);
-            double totalCost = calcFileCost(nodeGroup, sizeGB, fileAgeDays, numDiskWrites, numDiskReads);
-            file->setPropReal(getDFUQResultFieldName(DFUQRFcost), totalCost);
+            double atRestCost, accessCost;
+            calcFileCost(nodeGroup, sizeGB, fileAgeDays, numDiskWrites, numDiskReads, atRestCost, accessCost);
+            file->setPropReal(getDFUQResultFieldName(DFUQRFcost), atRestCost+accessCost);
+            file->setPropReal(getDFUQResultFieldName(DFUQRFatRestCost), atRestCost);
+            file->setPropReal(getDFUQResultFieldName(DFUQRFaccessCost), accessCost);
         }
 
         IPropertyTree *deserializeFileAttr(MemoryBuffer &mb, StringArray& nodeGroupFilter)

--- a/dali/base/dadfs.hpp
+++ b/dali/base/dadfs.hpp
@@ -282,7 +282,9 @@ enum DFUQResultField
     DFUQRFcost = 24,
     DFUQRFnumDiskReads = 25,
     DFUQRFnumDiskWrites = 26,
-    DFUQRFterm = 27,
+    DFUQRFatRestCost = 27,
+    DFUQRFaccessCost = 28,
+    DFUQRFterm = 29,
     DFUQRFreverse = 256,
     DFUQRFnocase = 512,
     DFUQRFnumeric = 1024,
@@ -419,7 +421,7 @@ interface IDistributedFile: extends IInterface
     virtual bool getSkewInfo(unsigned &maxSkew, unsigned &minSkew, unsigned &maxSkewPart, unsigned &minSkewPart, bool calculateIfMissing) = 0;
     virtual int  getExpire() = 0;
     virtual void setExpire(int expireDays) = 0;
-    virtual double getCost(const char * cluster) = 0;
+    virtual void getCost(const char * cluster, double & atRestCost, double & accessCost) = 0;
 };
 
 
@@ -855,7 +857,8 @@ inline const char *queryFileKind(IFileDescriptor *f) { return queryFileKind(f->q
 extern da_decl void ensureFileScope(const CDfsLogicalFileName &dlfn, unsigned timeoutms=INFINITE);
 
 extern da_decl bool checkLogicalName(const char *lfn,IUserDescriptor *user,bool readreq,bool createreq,bool allowquery,const char *specialnotallowedmsg);
-extern da_decl double calcFileCost(const char * cluster, double sizeGB, double fileAgeDays, __int64 numDiskWrites, __int64 numDiskReads);
+extern da_decl void calcFileCost(const char * cluster, double sizeGB, double fileAgeDays, __int64 numDiskWrites, __int64 numDiskReads, double & atRestCost, double & accessCost);
+extern da_decl double calcFileAccessCost(const char * cluster, __int64 numDiskWrites, __int64 numDiskReads);
 constexpr bool defaultPrivilegedUser = true;
 constexpr bool defaultNonPrivilegedUser = false;
 

--- a/dali/daliadmin/daliadmin.cpp
+++ b/dali/daliadmin/daliadmin.cpp
@@ -726,8 +726,10 @@ static void testDFSFile(IDistributedFile *legacyDfsFile, const char *logicalName
     PROGLOG("expire: %d", expire);
     try
     {
-        double cost = legacyDfsFile->getCost(clusterName.str());
-        PROGLOG("cost: %f", cost);
+        double atRestCost, accessCost;
+        legacyDfsFile->getCost(clusterName.str(), atRestCost, accessCost);
+        PROGLOG("accessCost: %f", accessCost);
+        PROGLOG("atRestCost: %f", atRestCost);
     }
     catch(IException *e)
     {

--- a/ecl/hthor/hthor.cpp
+++ b/ecl/hthor/hthor.cpp
@@ -761,7 +761,7 @@ void CHThorDiskWriteActivity::publish()
         {
             StringBuffer clusterName;
             file->getClusterName(0, clusterName);
-            diskAccessCost = money2cost_type(calcFileCost(clusterName, 0, 0, numDiskWrites, 0));
+            diskAccessCost = money2cost_type(calcFileAccessCost(clusterName, numDiskWrites, 0));
         }
         file->attach(logicalName.get(), agent.queryCodeContext()->queryUserDescriptor());
         agent.logFileAccess(file, "HThor", "CREATED", graph);
@@ -1357,7 +1357,7 @@ void CHThorIndexWriteActivity::execute()
 
         StringBuffer clusterName;
         dfile->getClusterName(0, clusterName);
-        diskAccessCost = money2cost_type(calcFileCost(clusterName, 0, 0, numDiskWrites, 0));
+        diskAccessCost = money2cost_type(calcFileAccessCost(clusterName, numDiskWrites, 0));
     }
     else
         lfn = filename;
@@ -8460,7 +8460,7 @@ void CHThorDiskReadBaseActivity::closepart()
                 dFile->addAttrValue("@numDiskReads", curDiskReads);
                 StringBuffer clusterName;
                 dFile->getClusterName(0, clusterName);
-                diskAccessCost += money2cost_type(calcFileCost(clusterName, 0, 0, 0, curDiskReads));
+                diskAccessCost = money2cost_type(calcFileAccessCost(clusterName, 0, curDiskReads));
             }
             numDiskReads += curDiskReads;
         }

--- a/esp/clients/ws_dfsclient/ws_dfsclient.cpp
+++ b/esp/clients/ws_dfsclient/ws_dfsclient.cpp
@@ -231,10 +231,8 @@ public:
     virtual bool isExternal() const override { return false; }
     virtual bool getSkewInfo(unsigned &maxSkew, unsigned &minSkew, unsigned &maxSkewPart, unsigned &minSkewPart, bool calculateIfMissing) override { return legacyDFSFile->getSkewInfo(maxSkew, minSkew, maxSkewPart, minSkewPart, calculateIfMissing); }
     virtual int getExpire() override { return legacyDFSFile->getExpire(); }
-    virtual double getCost(const char * cluster) override { return legacyDFSFile->getCost(cluster); }
-
-
-
+    virtual void getCost(const char * cluster, double & atRestCost, double & accessCost) override { legacyDFSFile->getCost(cluster, atRestCost, accessCost); }
+    
 // setters (change file meta data)
     virtual void setPreferredClusters(const char *clusters) override { legacyDFSFile->setPreferredClusters(clusters); }
     virtual void setSingleClusterOnly() override { legacyDFSFile->setSingleClusterOnly(); }

--- a/esp/scm/ws_dfu.ecm
+++ b/esp/scm/ws_dfu.ecm
@@ -191,7 +191,9 @@ ESPStruct [nil_remove] DFUFileDetail
     [min_ver("1.39")] ESParray<ESPstruct DFUFileBloom> Blooms;
     [min_ver("1.40")] int ExpireDays;
     [min_ver("1.41")] string KeyType;
-    [min_ver("1.60")] double Cost;
+    [min_ver("1.60"), max_ver("1.61")] double Cost;
+    [min_ver("1.62")] double AtRestCost;
+    [min_ver("1.62")] double AccessCost;
 };
 
 ESPStruct DFUSpaceItem
@@ -968,8 +970,8 @@ ESPresponse [exceptions_inline, nil_remove] DFUFilePublishResponse
 //  ===========================================================================
 ESPservice [
     auth_feature("DEFERRED"),
-    version("1.61"),
-    default_client_version("1.61"),
+    version("1.62"),
+    default_client_version("1.62"),
     noforms,
     exceptions_inline("./smc_xslt/exceptions.xslt")] WsDfu
 {

--- a/esp/scm/ws_dfu_common.ecm
+++ b/esp/scm/ws_dfu_common.ecm
@@ -52,5 +52,7 @@ ESPStruct DFULogicalFile
     [min_ver("1.41")] string KeyType;
     [min_ver("1.52")] int NumOfSubfiles;
     [min_ver("1.55")] string Accessed;
-    [min_ver("1.61")] double Cost;
+    [min_ver("1.61"), max_ver("1.61")] double Cost;
+    [min_ver("1.62")] double AtRestCost;
+    [min_ver("1.62")] double AccessCost;
 };

--- a/esp/services/ws_dfu/ws_dfuHelpers.cpp
+++ b/esp/services/ws_dfu/ws_dfuHelpers.cpp
@@ -184,7 +184,15 @@ bool WsDFUHelpers::addToLogicalFileList(IPropertyTree& file, const char* nodeGro
                 lFile->setAccessed(accessed.replace('T', ' '));
         }
         if (version >= 1.61)
-            lFile->setCost(file.getPropReal(getDFUQResultFieldName(DFUQRFcost)));
+        {
+            if (version < 1.62)
+                lFile->setCost(file.getPropReal(getDFUQResultFieldName(DFUQRFcost)));
+            else
+            {
+                lFile->setAtRestCost(file.getPropReal(getDFUQResultFieldName(DFUQRFatRestCost)));
+                lFile->setAccessCost(file.getPropReal(getDFUQResultFieldName(DFUQRFaccessCost)));
+            }
+        }
 
         logicalFiles.append(*lFile.getClear());
     }

--- a/esp/services/ws_dfu/ws_dfuService.cpp
+++ b/esp/services/ws_dfu/ws_dfuService.cpp
@@ -2655,7 +2655,18 @@ void CWsDfuEx::doGetFileDetails(IEspContext &context, IUserDescriptor *udesc, co
         }
     }
     if (version >= 1.60)
-        FileDetails.setCost(df->getCost(cluster));
+    {
+        double atRestCost, accessCost;
+        df->getCost(cluster, atRestCost, accessCost);
+
+        if (version <= 1.61)
+            FileDetails.setCost(atRestCost+accessCost);
+        else
+        {
+            FileDetails.setAccessCost(atRestCost);
+            FileDetails.setAtRestCost(accessCost);
+        }
+    }
     PROGLOG("doGetFileDetails: %s done", name);
 }
 

--- a/thorlcr/graph/thgraphmaster.cpp
+++ b/thorlcr/graph/thgraphmaster.cpp
@@ -562,7 +562,7 @@ void CMasterActivity::updateFileReadCostStats(std::vector<OwnedPtr<CThorStatsCol
                 stat_type numDiskReads = subFileStats[i]->getStatisticSum(StNumDiskReads);
                 StringBuffer clusterName;
                 file->getClusterName(0, clusterName);
-                diskAccessCost += money2cost_type(calcFileCost(clusterName, 0, 0, 0, numDiskReads));
+                diskAccessCost += money2cost_type(calcFileAccessCost(clusterName, 0, numDiskReads));
                 file->addAttrValue("@numDiskReads", numDiskReads);
             }
         }
@@ -575,7 +575,7 @@ void CMasterActivity::updateFileReadCostStats(std::vector<OwnedPtr<CThorStatsCol
             stat_type numDiskReads = statsCollection.getStatisticSum(StNumDiskReads);
             StringBuffer clusterName;
             file->getClusterName(0, clusterName);
-            diskAccessCost += money2cost_type(calcFileCost(clusterName, 0, 0, 0, numDiskReads));
+            diskAccessCost += money2cost_type(calcFileAccessCost(clusterName, 0, numDiskReads));
             file->addAttrValue("@numDiskReads", numDiskReads);
         }
     }
@@ -589,7 +589,7 @@ void CMasterActivity::updateFileWriteCostStats(IFileDescriptor & fileDesc, IProp
         assertex(fileDesc.numClusters()>=1);
         StringBuffer clusterName;
         fileDesc.getClusterGroupName(0, clusterName);// Note: calculating for 1st cluster. (Future: calc for >1 clusters)
-        diskAccessCost += money2cost_type(calcFileCost(clusterName, 0, 0, numDiskWrites, 0));
+        diskAccessCost = money2cost_type(calcFileAccessCost(clusterName, numDiskWrites, 0));
     }
 }
 

